src/java/main/org/apache/zookeeper/server/NettyServerCnxn.java

```diff
@@ -87,6 +87,12 @@ public void close() {
             LOG.debug("close called for sessionid:0x"
                     + Long.toHexString(sessionId));
         }
+
+        // ZOOKEEPER-2743:
+        // Always unregister connection upon close to prevent
+        // connection bean leak under certain race conditions.
+        factory.unregisterConnection(this);
```

Comment:
>> Could you please raise a new jira to track the changes.

Created https://issues.apache.org/jira/browse/ZOOKEEPER-2751 for this. Merging this PR now.